_SPIKE_.md

```diff
@@ -0,0 +1,496 @@
+# Spike: Resolve included models
+
+Consider the following domain model(s):
+
+- A model called `Category` with properties `id`, `name`.
+- A model called `Product` with properties `id`, `name`, `categoryId`
+- A `hasMany` relation (Category has many products)
+
+Now consider the following code to retrieve all categories with all related
+products (perhaps to render a home page of a product catalog):
+
+```ts
+categoryRepo.find({include: [{relation: 'products'}]});
+```
+
+## The problem
+
+How to fetch products for each category found?
+
+Additional constraints:
+
+- The target model (`Product`) can be backed by a different database than the
+  source model (`Category`). For example, we can use MongoDB to store categories
+  and MySQL to store products.
+
+- We need to test relations against real databases to ensure we are correctly
+  handling database-specific quirks like:
+  - a limit on the number of items that can be included in `inq`
+  - coercion between `ObjectID` vs. `string` for MongoDB.
+
+## Proposed solution
+
+### 1. Introduce the concept of `InclusionResolver`
+
+An inclusion resolver is a class that can fetch target models for the given list
+of source model instances. The idea is to create a different inclusion resolver
+for each relation type, e.g. `HasManyInclusionResolver`, etc.
+
+```ts
+export interface InclusionResolver {
+  fetchIncludedModels<SourceWithRelations extends Entity>(
+    // list of source models as returned by the first database query
+    sourceEntities: SourceWithRelations[],
+    // inclusion requested by the user (e.g. scope constraints to apply)
+    inclusion: Inclusion,
+    // generic options object, e.g. carrying the Transaction object
+    options?: Options,
```

Comment:
Is it possible to retrieve metadata describing the relation?
Do we want to introduce `InclusionResolutionContext`?

Reply:
Metadata describing the relation is always the same for every resolver invocation, therefore it's not passed in the arguments. Instead, the metadata is provided to the class constructor. Similarly how a getter of a repository for the target model is provided.

> Do we want to introduce `InclusionResolutionContext`

I am not sure.

On the one hand, the current API where `inclusion` and `options` are provided via positional arguments, is simple and easy to understand. `InclusionResolutionContext` feels rather heavy-weight to me.

On the other hand, positional parameters are difficult to extend in the future. `InclusionResolutionContext` makes it easy to add additional context properties.

If we introduce `InclusionResolutionContext`, what properties would you like to see on that context object? Just `inclusion` and `options`?